### PR TITLE
Geneformer modelcard and loss eval script

### PR DESCRIPTION
**Pull request by jstjohn**
## Summary
Port the geneformer loss eval script over from bionemo1, and update the model card with new numbers and results.

### CODE REVIEW

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            self.nv_tokenizer.sep_token: my_token_dict["<eos>"],  # name doesn't really matter here
+        }
+        tokens = list(my_token_dict.items())
+        for k, t in tokens[:4]:
+            assert k.startswith("<")
+
+        missing_nv_tokens = []
+        extra_tokens_not_covered = []
+        for ens, idx in list(my_token_dict.items())[4:]:
+            assert ens.startswith("ENSG")
```
Comment (polinabinder1) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:82]: Where is ENSG coming from?

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    token_dictionary_path: Path,
+    mask_prob: float = 0.15,
+    batch_size: int = 16,
+    precision: str = "bf16-mixed",
+    config_class: Type[BioBertConfig] = GeneformerConfig,
+    seed: int = 513,
+):
+    """Inference function (requires DDP and only training data that fits in memory)."""
+    # This is just used to get the tokenizer :(
+    train_data_path: Path = (
```
Comment (polinabinder1) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:130]: Why do we only want this training data path?

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py (hunk 1 above)
Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:82]: This is a sanity check that the user is pointing their data at ENSG (ensembl) format gene ids.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py (hunk 2 above)
Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:130]: This has the tokenizer in it.

File: docs/docs/assets/images/geneformer/loss_curve_new_v_old_geneformer_64_node_10M.png
[the remark this answers is not in the document]
Comment (jstjohn) [reply docs/docs/assets/images/geneformer/loss_curve_new_v_old_geneformer_64_node_10M.png:1]: I just downloaded from wandb.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+from bionemo.geneformer.data.singlecell.preprocess import GeneformerPreprocess
+from bionemo.geneformer.tokenizer.gene_tokenizer import GeneTokenizer
+from bionemo.llm.data import collate
+from bionemo.llm.model.biobert.model import BioBertConfig
+from bionemo.testing import megatron_parallel_state_utils
+from bionemo.testing.data.load import load
+
+
+class GeneformerHFAdapter(torch.nn.Module):
+    """An adapter class for running the HF model against our subset of tokens."""
```
[the remark this answers is not in the document]
Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:58]: This is only storing mappings of the tokenizer. This doesn't have a constraint that the dataset must fit into memory.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:1]: Unfortunately there are two manual download steps documented in the `--help` messsage.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py (hunk 1 above)
Comment (polinabinder1) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:83]: I misunderstood from a comment lower down.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py (hunk 2 above)
Comment (polinabinder1) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:134]: Would this work with another tokenizer? If not, I can see the training path download + test case issue. But in that case, couldn't we have a test case that also downloads this data from PBSS to that location?

Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:134]: This script is very one off for this purpose. I was not trying to make something too general. We can update it in the future to make it compatible with a future version of geneformer that uses a different tokenizer. For now though my goal was to port this script over from bionemo1 and update the numbers in the modelcard.

File: docs/docs/models/geneformer.md [hunk 5]
```diff
@@ -166,19 +177,39 @@ This checkpoint was trained for approximately 11 epochs through the CELLxGENE sp
...
+    directly. Some key differences follow:
+
+     * Trained on a much larger 95M cell dataset. Our current checkpoints were trained with 23M cells.
+     * The new 12 layer baseline geneformer variant sits between our 10M and 106M parameter models in parameter count with
+      approximately 38M parameters.
+     * The model is trained with a 4096 context rather than a 2048 context. When forcing the model to make predictions
+      with a 2048 context, the MLM loss drops to *2.76*, which is probably unfair because this may be "out of domain" for
+      training. It is really hard to compare these loss numbers directly is the only take-home here.
+     * The model was trained on a set of 20275 genes, rather than the older set of 25426 genes. This would also be
+      expected to give a boost in loss since there are fewer tokens to chose from.
```
Comment (pstjohn) [inline docs/docs/models/geneformer.md:212]:
```suggestion
     * The model was trained on a set of 20,275 genes, rather than the older set of 25,426 genes. This would also be
      expected to give a boost in loss since there are fewer tokens to choose from.
```

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        required=True,
+    )
+    parser.add_argument(
+        "--hf-medians-dictionary-path",
+        type=Path,
+        help="Path to token dictionary file. "
+        "Eg `wget https://huggingface.co/ctheodoris/Geneformer/resolve/main/geneformer/gene_median_dictionary_gc95M.pkl` "
+        "then provide the path to the downloaded file.",
+        required=True,
+    )
```
Comment (pstjohn) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:299]: Any reason you couldn't just use `pooch` directly and point it to these files? You could save them to `BIONEMO_CACHE_DIR` and it should be easier to put a test around this

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        download_directory=train_data_path,
+        medians_file_path=train_data_path / "medians.json",
+        tokenizer_vocab_path=train_data_path / "geneformer.vocab",
+    )
+    match preprocessor.preprocess():
+        case {"tokenizer": tokenizer, "median_dict": median_dict}:
+            logging.info("*************** Preprocessing Finished ************")
+        case _:
+            logging.error("Preprocessing failed.")
+            assert False
```
Comment (pstjohn) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:149]: Maybe raise a more informative error here?

**Review by pstjohn — APPROVED** [review]

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:1]: I made an issue for this #401

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py (hunk 6 above)
Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:299]: I made an issue for this #401

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py (hunk 2 above)
Comment (jstjohn) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/geneformer_mlm_loss_eval.py:134]: Created an issue for this #402

**Review by polinabinder1 — APPROVED** [review]

**Review by skothenhill-nv — APPROVED** [review]
